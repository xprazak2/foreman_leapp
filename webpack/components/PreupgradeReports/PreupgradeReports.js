import React from 'react';
import MessageBox from 'foremanReact/components/common/MessageBox';
import { LoadingState, Row } from 'patternfly-react';
import PropTypes from 'prop-types';

import PreupgradeReportsList from '../PreupgradeReportsList';
import FixAllButton from './components/FixAllButton';

import {
  flattenEntries,
  isEmpty,
  anyEntriesFixable,
} from './PreupgradeReportsHelpers';

const PreupgradeReports = ({
  preupgradeReports,
  loading,
  error,
  csrfToken,
  newJobInvocationUrl,
}) => {
  if (!isEmpty(error)) {
    return (
      <MessageBox
        key="preupgrade-reports-error"
        icontype="error-circle-o"
        msg={`Could not retrieve data: ${error.statusText} - ${error.errorMsg}`}
      />
    );
  }

  return (
    <LoadingState loading={loading}>
      <Row>
        <div className="title-filter col-md-4">&nbsp;</div>
        <div id="title_action" className="col-md-8">
          <div className="btn-toolbar pull-right">
            <FixAllButton
              postUrl={newJobInvocationUrl}
              disabled={!anyEntriesFixable(preupgradeReports)}
              csrfToken={csrfToken}
              preupgradeReports={preupgradeReports}
            />
          </div>
        </div>
      </Row>
      <PreupgradeReportsList allEntries={flattenEntries(preupgradeReports)} />
    </LoadingState>
  );
};

PreupgradeReports.propTypes = {
  preupgradeReports: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.object.isRequired,
  csrfToken: PropTypes.string.isRequired,
  newJobInvocationUrl: PropTypes.string.isRequired,
};

export default PreupgradeReports;
