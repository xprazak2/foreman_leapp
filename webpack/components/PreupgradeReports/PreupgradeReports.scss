@import '~@theforeman/vendor/scss/variables';

#leapp_preupgrade_report {
  .list-view-pf.working {
    .list-group-item.list-view-pf-stacked {
      color: $color-pf-black-400;

      &:hover{
        background-color: $color-pf-white;
        border-left-color: $color-pf-white;
        border-right-color: $color-pf-white;
      }
    }
  }
}
