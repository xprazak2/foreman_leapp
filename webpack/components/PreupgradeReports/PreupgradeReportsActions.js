import api from 'foremanReact/API';
import { deepPropsToCamelCase } from 'foremanReact/common/helpers';

import {
  PREUPGRADE_REPORTS_REQUEST,
  PREUPGRADE_REPORTS_SUCCESS,
  PREUPGRADE_REPORTS_FAILURE,
} from '../../consts';

export const getPreupgradeReports = url => async dispatch => {
  dispatch({ type: PREUPGRADE_REPORTS_REQUEST });

  try {
    const { data } = await api.get(url);
    return dispatch({
      type: PREUPGRADE_REPORTS_SUCCESS,
      payload: deepPropsToCamelCase(data),
    });
  } catch (error) {
    return dispatch(errorHandler(PREUPGRADE_REPORTS_FAILURE, error));
  }
};

const errorHandler = (msg, err) => {
  const error = {
    errorMsg: 'Failed to fetch preupgrade reports from server.',
    statusText: err.response.statusText,
  };
  return { type: msg, payload: { error } };
};
