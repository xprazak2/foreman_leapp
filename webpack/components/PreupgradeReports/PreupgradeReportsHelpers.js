export const flattenEntries = reports =>
  reports.reduce((memo, report) => [...memo, ...report.entries], []);

export const entryFixable = entry =>
  entry.detail &&
  entry.detail.remediations &&
  entry.detail.remediations.some(remediation => remediation.type === 'command');

export const isEmpty = obj => Object.keys(obj).length === 0;

export const anyEntriesFixable = reports =>
  flattenEntries(reports).some(entryFixable);

export const idsForInvocation = reports =>
  reports.reduce(
    (memo, report) => {
      report.entries.forEach(entry => {
        if (entryFixable(entry)) {
          memo.entryIds = [...memo.entryIds, entry.id];

          if (!memo.hostIds.includes(report.hostId)) {
            memo.hostIds = [...memo.hostIds, report.hostId];
          }
        }
      });
      return memo;
    },
    { hostIds: [], entryIds: [] }
  );
