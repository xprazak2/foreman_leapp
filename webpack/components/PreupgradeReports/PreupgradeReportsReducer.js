import Immutable from 'seamless-immutable';

import {
  PREUPGRADE_REPORTS_REQUEST,
  PREUPGRADE_REPORTS_SUCCESS,
  PREUPGRADE_REPORTS_FAILURE,
  PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_REQUEST,
  PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_SUCCESS,
  PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_FAILURE,
} from '../../consts';

export const initialState = Immutable({
  loadingPreupgradeReports: false,
  preupgradeReports: [],
  error: {},
  fixAllWorking: false,
  fixAllError: {},
});

const reducer = (state = initialState, action) => {
  const { payload } = action;
  switch (action.type) {
    case PREUPGRADE_REPORTS_REQUEST:
      return state.set('loadingPreupgradeReports', true);
    case PREUPGRADE_REPORTS_SUCCESS:
      return state.merge({
        loadingPreupgradeReports: false,
        preupgradeReports: payload,
      });
    case PREUPGRADE_REPORTS_FAILURE:
      return state.merge({
        error: payload.error,
        loadingPreupgradeReports: false,
      });
    case PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_REQUEST:
      return state.set('fixAllWorking', true);
    case PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_SUCCESS:
      return state.set('fixAllWorking', false);
    case PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_FAILURE:
      return state.merge({
        fixAllWorking: false,
        fixAllError: { ...payload }, // what is payload???!
      });
    default:
      return state;
  }
};

export default reducer;
