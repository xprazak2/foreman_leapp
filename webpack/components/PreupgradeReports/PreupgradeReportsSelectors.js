export const selectPreupgrade = state => state.foremanLeapp.preupgrade;

export const selectPreupgradeReports = state => selectPreupgrade(state).preupgradeReports;
export const selectLoadingPreupgradeReports = state =>
  selectPreupgrade(state).loadingPreupgradeReports;
export const selectError = state => selectPreupgrade(state).error;
