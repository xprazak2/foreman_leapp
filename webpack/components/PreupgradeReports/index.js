import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import PreupgradeReports from './PreupgradeReports';

import getCsrfToken from '../../csrf';

import {
  selectPreupgradeReports,
  selectLoadingPreupgradeReports,
  selectError,
} from './PreupgradeReportsSelectors';

import { getPreupgradeReports } from './PreupgradeReportsActions';

const WrappedPreupgradeReports = ({ url, newJobInvocationUrl }) => {
  const loading = useSelector(state => selectLoadingPreupgradeReports(state));
  const preupgradeReports = useSelector(state =>
    selectPreupgradeReports(state)
  );
  const error = useSelector(state => selectError(state));

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getPreupgradeReports(url));
  }, [url, dispatch]);

  return (
    <PreupgradeReports
      preupgradeReports={preupgradeReports}
      error={error}
      loading={loading}
      csrfToken={getCsrfToken()}
      newJobInvocationUrl={newJobInvocationUrl}
    />
  );
};

WrappedPreupgradeReports.propTypes = {
  url: PropTypes.string.isRequired,
  newJobInvocationUrl: PropTypes.string.isRequired,
};

export default WrappedPreupgradeReports;
