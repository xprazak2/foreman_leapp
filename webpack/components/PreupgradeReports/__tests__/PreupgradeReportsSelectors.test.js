import { testSelectorsSnapshotWithFixtures } from '@theforeman/test';

import { preupgradeReports } from './PreupgradeReports.fixtures';

import {
  selectPreupgradeReports,
  selectLoadingPreupgradeReports,
  selectError,
} from '../PreupgradeReportsSelectors';

const state = {
  foremanLeapp: {
    preupgrade: {
      preupgradeReports,
      loadingPreupgradeReports: false,
      error: { errorMsg: 'Error', statusText: 'Status text' },
    },
  },
};

const fixtures = {
  'should return preupgrade reports': () => selectPreupgradeReports(state),
  'should return loading state': () => selectLoadingPreupgradeReports(state),
  'should return error': () => selectError(state),
};

describe('PreupgradeReportsSelectors', () =>
  testSelectorsSnapshotWithFixtures(fixtures));
