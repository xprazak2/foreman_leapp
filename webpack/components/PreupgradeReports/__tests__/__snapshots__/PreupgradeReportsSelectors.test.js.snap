// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreupgradeReportsSelectors should return error 1`] = `
Object {
  "errorMsg": "Error",
  "statusText": "Status text",
}
`;

exports[`PreupgradeReportsSelectors should return loading state 1`] = `false`;

exports[`PreupgradeReportsSelectors should return preupgrade reports 1`] = `
Array [
  Object {
    "entries": Array [
      Object {
        "id": 42,
        "severity": "Too severe to talk about",
        "title": "Fix me!",
      },
      Object {
        "id": 43,
        "severity": "medium",
        "title": "I am broken too",
      },
    ],
    "hostId": 5,
  },
  Object {
    "entries": Array [
      Object {
        "id": 44,
        "severity": "high",
        "title": "Octocat is not happy",
      },
      Object {
        "id": 45,
        "severity": "low",
        "title": "Not enough credits",
      },
    ],
    "hostId": 6,
  },
]
`;
